plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TemporalTimeTravelUtil.java

```diff
@@ -0,0 +1,170 @@
+        if (checkpointEntries.isEmpty()) {
+            throw new MissingTransactionLogException(getTransactionLogJsonEntryPath(transactionLogDir, entryNumber).toString());
+        }
+
+        Optional<CommitInfoEntry> commitInfo = checkpointEntries.get().getEntries(fileSystem)
+                .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                .filter(Objects::nonNull)
+                .findFirst();
+
+        // We can not derive useful info from the last check point
+        if (commitInfo.isEmpty()) {
+            return findLatestVersionFromWholeTransactions(fileSystem, tableLocation, epochMillis);
+        }
+
+        if (commitInfo.get().timestamp() == epochMillis) {
+            return entryNumber;
+        }
+
+        if (commitInfo.get().timestamp() < epochMillis) {
+            long tail = searchTowardsTail(fileSystem, entryNumber + 1, transactionLogDir, epochMillis);
+            if (tail >= 0) {
+                return tail;
+            }
+            return entryNumber;
+        }
+        else {
+            long head = searchTowardsHead(fileSystem, entryNumber - 1, transactionLogDir, epochMillis);
+            if (head >= 0) {
+                return head;
+            }
+        }
+
+        throw new TrinoException(INVALID_ARGUMENTS, format("No temporal version history at or before %s", Instant.ofEpochMilli(epochMillis)));
+    }
+
+    private static long searchTowardsTail(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis)
+            throws IOException
+    {
+        long version = -1;
+        Optional<TransactionLogEntries> entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        while (entries.isPresent()) {
+            Optional<CommitInfoEntry> commitInfo = entries.get().getEntries(fileSystem)
+                    .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                    .filter(Objects::nonNull)
+                    .findFirst();
+
+            if (commitInfo.isEmpty()) {
+                entryNumber++;
+                entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+                continue;
+            }
+
+            if (commitInfo.get().timestamp() > epochMillis) {
+                break;
+            }
+
+            version = entryNumber;
+
+            entryNumber++;
+            entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        }
+
+        return version;
+    }
+
+    private static long searchTowardsHead(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis)
+            throws IOException
+    {
+        Optional<TransactionLogEntries> entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
```

Comment:
You don't need to read ALL the entries from JSON, but rather the CommitInfo.

Take inspiration from `io.trino.plugin.deltalake.transactionlog.TransactionLogEntries#getMetadataAndProtocol`

Reply:
The `getEntriesFromJson` not read all the entries, the reading is actual happening at the call of the `TransactionLogEntries#getEntries`, fortunately it returns stream, and here use the findFirst, it's will break if meet the condition in advance.